Add support to total batch size argument for transformers transfer learning

anmarques opened the pull request:
Created the NMTrainingArguments that inherits from HF's TrainingArguments. This class allows one to add arguments to the training script and handling potential conflicts with other arguments. In particular, added "total_train_batch_size" and "total_eval_batch_size" and made sure these arguments cannot be used alongside "per_device_train_batch_size" and "per_device_eval_batch_size."

Commit:
Added NMTrainingArgument class, and extension of HF's TrainingArgument class. Included support to total batch size arguments.

Commit:
Added NMTrainingArgument class, and extension of HF's TrainingArgument class. Included support to total batch size arguments.

Commit:
Added NMTrainingArgument class, and extension of HF's TrainingArgument class. Included support to total batch size arguments.

Commit:
Added NMTrainingArgument class, and extension of HF's TrainingArgument class. Included support to total batch size arguments.

Commit:
Fixed relative path to NMTrainingArguments

Commit:
Fixed relative path to NMTrainingArguments

Commit:
Added support to total batch size argument to masked lamguage modeling, question_answering and text classification

Commit:
Merge branch 'qa-batchsize' of github.com:anmarques/sparseml into qa-batchsize

Commit:
Removed double import of NMTrainingArguments

rahul-tuli reviewed (dismissed):
<img src="https://media2.giphy.com/media/gVoBC0SuaHStq/giphy.gif"/>

KSGulin committed:
Update base pruning args (#559)

* Update: add explicit kwargs passing to avoid multiple inheritance issue

bfineran committed:
[MFAC] Small block optimized FisherInverse implementation (#487)

* [MFAC] Small block optimized FisherInverse implementation
* [MFAC] training aware implementation for GradSampler
* [MFAC] Integration Into BERT
* [MFAC] Updated flow to use small block version in all reasonable cases

rahul-tuli committed:
Activation Bits Support + qconfig kwargs for QuantizationModifier (#542)

* Add: Support for 4 bit activation quantization

* Apply consistent typing for `activation_kwargs`

* observer quant_min/max bugfix

* Add: Support for `weight_qconfig_kwargs`
Fix: tests, test helper names, style
Fix - Merge Conflicts
Renamed - `enable_in4_activations` to `int4_activations`

* Address: Test failures
Use: A copy of `activation_qconfig_kwargs` for overriding quant_max, quant_min values

* Update: To a more general approach, using `activation_bits` instead of hardcoding to 4 bits

* Addressed review  comments from bfineran

* Style

* Addressed: Conv3d rebase

* Add: Tests for `activation_qconfig_kwargs`

* Addressed: Review comments from @natuan

* Revert "Addressed: Review comments from @natuan"

This reverts commit 7f6ac86650c5dcdc162d385d29c1d7544d9e8b2c.

Co-authored-by: Benjamin <ben@neuralmagic.com>

KSGulin committed:
Pruning modifier testing with init kwargs (#565)

* Update: support for pruning modifiers with additional init kwargs

natuan commented on src/sparseml/transformers/nm_training_arguments.py:
```diff
@@ -0,0 +1,96 @@
+        # If total batch size is defined at this point,
+        # per device batch size needs to be computed
+        if self.total_train_batch_size:
+            self.per_device_train_batch_size = self.total_train_batch_size // max(
```
Would it be better if we use this total_train_batch_size to dictate both the per_device_train_batch_size and the gradient_accumulation_steps (so the "steps" here mean backward steps)---this way we make sure that this total batch size, if specified, is consistent even between runs with different num gpus? We could start with an accumulation step of 1, calculating the corresponding per device batch size; if that fails then we double the accumulation step (and so on).

anmarques replied:
Hi Tuan. I fail to see how gradient_accumulation_steps can be set "consistently" from the total batch size alone. In my understanding, the accumulation batch size will be total_train_batch_size * gradient_accumulation_steps. On the other hand, per_device_train_batch_size = total_batch_size / nGPUs, regardless of the accumulation steps.

natuan replied:
Hi Alex, my thinking was that we could interpret "total_train_batch_size" as your "accumulation batch size" above, which I think is consistent with the base trainer code here: https://github.com/neuralmagic/transformers/blob/master/src/transformers/trainer.py#L1085
in which the args.train_batch_size I think is computed from per_device_batch_size and num gpus as here: https://github.com/neuralmagic/transformers/blob/master/src/transformers/training_args.py#L699

So taking all together, we would have total_train_batch_size is a func of per_device_batch_size, gradient_accumulation_steps, and num gpus. We know num gpus, and then I'm thinking we can adjust the other two params for the batch to fit gpu memory.

anmarques replied:
I think I understand that now. However, how would we check if it the batch fits into GPU memory? I would rather allow the user to specify either total_train_batch_size or per_device_train_batch_size (together with gradient_accumulation_steps) and compute the other from the expression in L1085 in trainer.py above. Does that address your concern?

anmarques replied:
Please review the latest version. This one should have the total batch size take the accumulation steps into consideration

bfineran committed:
add gputils to torch deps (#567)

* add gputils to torch deps

* quality

rahul-tuli committed:
Add: `sparseml.image_classification` pathways (#566)

* Add: `sparseml.image_classification.train` and `sparseml.image_classification.export_onnx` pathways

* Remove: ~test_nmargparser.py~ from `src`

* Style fixes

* - [X] Addressed review comments
- [X] Deleted scripts from integration
- [X] Added lr_analysis pathway
- [X] Added pr_sensitivity pathway
- [X] Moved production code to `sparseml.pytorch.image_classification`
- [X] Moved helpers to utils folder

* Addressed review comments from @bfineran

* Added tests for `nm_argparser.py`

* Updated: new script locations in `vision.py`

* Style

Co-authored-by: Benjamin Fineran <bfineran@users.noreply.github.com>

jeanniefinks committed:
Update README.md, index.rst, CONTRIB, CONDUCT, misc overview pages (#568)

* Update README.md

- updating Contributor Covenant 2.0 to 2.1
- retiring Discourse links
- elevating punctuation so they're within links vs outside for better visual

* Update CODE_OF_CONDUCT.md

 updating Contributor Covenant 2.0 to 2.1

* Update index.rst

Retiring Discourse

* Update installation.md

elevating punctuation so they're within links vs outside for better visual

* Update recipes.md

elevating punctuation so they're within links vs outside for better visual

* Update onnx_export.md

elevating punctuation so they're within links vs outside for better visual

* Update README.md

* Update CONTRIBUTING.md

dbogunowicz committed:
Implementation AC/DC (moment buffer reset & finish training on decompression epoch) (#544)

* Squashed commit of the following:

commit a6f5b5806305098997a08cfcb6d6d67c47589def
Author: bogunowicz@arrival.com <bogunowicz@arrival.com>
Date:   Mon Feb 7 17:52:17 2022 +0100

    woops

commit 2ed901cc9f0b87cdd420199a5388c4fd71b03fb0
Merge: 05125ee05f 867c7c57da
Author: bogunowicz@arrival.com <bogunowicz@arrival.com>
Date:   Mon Feb 7 17:48:47 2022 +0100

    Merge branch 'feature/ac_dc' of https://github.com/neuralmagic/sparseml into feature/ac_dc

commit 05125ee05f1cfd7d83636e7347fb03adee7add85
Author: bogunowicz@arrival.com <bogunowicz@arrival.com>
Date:   Wed Feb 2 16:06:01 2022 +0100

    ac_dc implementation ready, now testing

commit 867c7c57da4c3f105723aa8c9326e07c5aa903ae
Author: bogunowicz@arrival.com <bogunowicz@arrival.com>
Date:   Mon Feb 7 17:45:54 2022 +0100

    remove some local files

commit a490506503b1f568230026a59fb4fd55121491b9
Author: bogunowicz@arrival.com <bogunowicz@arrival.com>
Date:   Mon Feb 7 17:42:06 2022 +0100

    fixed style and quality

commit 3ffa83647aa3e54bdf7011cd397ea70c9d6768c4
Author: bogunowicz@arrival.com <bogunowicz@arrival.com>
Date:   Wed Feb 2 16:06:01 2022 +0100

    ac_dc implementation ready, now testing

* ready to test mobilenet on imagenet

* fixed minor blunder

* working code, ready to refactor

* correction according to Ben's hints, enabling glboal sparsity

* starting writing test for acdc

* initial refactoring on top of Ben's refactor PR

* .

* ACDCPruningModifier as a child of BasePruningModifier + recipe for completeness

* making recipe more compact

* clean up legacy files

* first fixes up to -> object has no attribute '_initialized'

* some progress with tests

* refactor tests

* all tests working

* Update src/sparseml/pytorch/sparsification/pruning/modifier_pruning_acdc.py

Co-authored-by: Benjamin Fineran <bfineran@users.noreply.github.com>

* small improvement to tests

* fixed missing logging

* add docstring

* Update modifier_pruning_acdc.py

* fix  inheritance

* leave_enabled conundrum fixed once for all

* yaml test fixes

* quality

* leave enabled fix

Co-authored-by: bogunowicz@arrival.com <bogunowicz@arrival.com>
Co-authored-by: Benjamin Fineran <bfineran@users.noreply.github.com>
Co-authored-by: Benjamin <ben@neuralmagic.com>
Co-authored-by: Mark Kurtz <mark@neuralmagic.com>

KSGulin committed:
MFAC Refactor (#558)

* M-FAC refactor to comply with pruning modifier refactor, along with smaller-scale updates to MFAC implementation and testing

bfineran committed:
stages support for Manager lifecycle (#562)

* stages support for Manager lifecycle

* load staged recipe yaml string into list of dicts

* test case fix

* testing for load staged recipe

* use OrderedDict + comparator key fix

* manager staged serialization test

* quality fix

* finalize and save structured sub-recipe

* staged recipe manager composition

Commit:
Adjusted total train batch size definition to include gradient accumulation steps.

bfineran committed:
StructuredPruningModifier Refactor (#560)

* StructuredPruningModifier Refactor

* migrate packages

* style

* nit fix

* rebase test fixes

* rebase fixes + mask type for mfac

* Fix: test MFAC mask type

* Update src/sparseml/pytorch/sparsification/pruning/modifier_pruning_structured.py

* style

rahul-tuli committed:
Quantization Code Move (#570)

* Moved:
- `modifier_quantization.py` to `sparseml.pytrorch.sparsification.quantization`
- Updated test location and usages
_ Style

* fix imports

* Style

* Moved `sparseml/pytorch/utils/quantization` to `sparseml/pytorch/sparsification/quantization/utils`

* Moved `sparseml/pytorch/utils/quantization` to `sparseml/pytorch/sparsification/quantization/utils`

* Style

* Style

* fix imports

* Style

* Review Comments

* fix quality + failing tests + cyclic dependency

* quality

Co-authored-by: Benjamin Fineran <bfineran@users.noreply.github.com>
Co-authored-by: Benjamin <ben@neuralmagic.com>

bfineran committed:
wandb logger test fix (#572)

bfineran committed:
move remaining torch pruning modifiers to /sparsification/pruning (#571)

* refactor movement and layer pruning modifiers + delete optim/modifier_pruning

* refactor note

* quality

* review fixes

* remove buggy test - to be fixed post release

KSGulin committed:
Fix mfac gradcheck (#573)

* Fix: count explicitly number of gradients collected

bfineran committed:
re-adding deprecated IC examples (#574)

dbogunowicz commented on src/sparseml/transformers/nm_training_arguments.py:
```diff
@@ -0,0 +1,102 @@
+        # If neither total nor per device batch sizes are defined
+        # set total batch sizes as default
+        if not self.total_train_batch_size and not self.per_device_train_batch_size:
+            self.total_train_batch_size = 8
```
I understand that `8` is the original, default value for both `total_train_batch_size` and `total_eval_batch_size`? I am concerned that we are somewhat "hiding" those default values from the user. But if user expectation is to obtain default value of `8` in HF TrainingArguments, I am probably cool with this.

anmarques replied:
I just used the default values from HF.

dbogunowicz commented on src/sparseml/transformers/nm_training_arguments.py:
```diff
@@ -0,0 +1,102 @@
+
+
+@dataclass
+class NMTrainingArguments(TrainingArguments):
```
Thank you for very clear walkthrough of the diff!

dbogunowicz commented on src/sparseml/transformers/nm_training_arguments.py:
```diff
@@ -0,0 +1,102 @@
+# Copyright (c) 2021 - present / Neuralmagic, Inc. All Rights Reserved.
```
Very general comment, is there a reason for forking the SparseML repository, creating branch locally in your fork, and then finally merging the PR into original repository's master?
I guess that for people who are inside NM (not external contributors), it would be advisable to work directly within our repo, since then other team members can e.g. quickly merge your branch or test your branch without cloning your fork. Or maybe there are some hidden benefits of your method which I am just failing to see? :)

anmarques replied:
I was just following the guidelines for contributions, no strong reason. But noted, will work from our repo for following changes.

KSGulin committed:
Fix: proper handeling of dictionary type num_grads for MFAC (#579)

KSGulin committed:
Fix: change QATConv3D instance to Conv3D (#578)

rahul-tuli committed:
Update trainer.py (#581)

Fix: Small typo in `RecipeManagerTrainerInterface` docstring

KSGulin committed:
M-FAC Indexing fix (#583)

* Fix: MFAC indexing bug on fisher mul

* Update: mfac index var names

* Fix: reduce number of runs for fisher inverse tests

KSGulin committed:
Update: docs to use correct mfac options format (#584)

dbogunowicz committed:
Ability to load recipe stages from YAML into dictionary of name -> List[modifiers] (#556)

* first proposal of the solution

* add docstring

* stages support for Manager lifecycle

* load staged recipe yaml string into list of dicts

* modifiers properly read sample staged recipe

* modifiers properly read sample staged recipe

* edit after Ben's comments

* tiny docstring fix

* Removed stage_recipy.md from repository

* Add fixes according to Ben's commments

* Nit: remove print functions

* Give priority to local variables over global when conflicting

* Allow information from main_container flow to the evaluation of stage_containers

Co-authored-by: bogunowicz@arrival.com <bogunowicz@arrival.com>
Co-authored-by: Benjamin <ben@neuralmagic.com>

dbogunowicz committed:
Update SparseML to support Pytorch 1.10.2 (#569)

* rebase on top of current main

* Make failing ONNX accept test cases for legacy or upgraded Python versions

* Fix quality and make sure we are upgrading to PyTorch 1.10.2 and not 1.10.0

* Update src/sparseml/pytorch/base.py

Co-authored-by: Rahul Tuli <rahul@neuralmagic.com>

Co-authored-by: bogunowicz@arrival.com <bogunowicz@arrival.com>
Co-authored-by: Benjamin Fineran <bfineran@users.noreply.github.com>
Co-authored-by: Rahul Tuli <rahul@neuralmagic.com>

KSGulin committed:
Mask creator int -> round (#589)

* Fix: round for pruning index to match _threshold_from_sparsity

rahul-tuli committed:
Remove: `export` to `onnx` by default (#591)

* Remove: `export` to `onnx` by default

* Remove: not needed params

Commit:
Added logging messages.

Commit:
Fixed typo.

Commit:
Reintroduce import of NMTrainingArguments.

spacemanidol commented:
@alexm-nm given how many commits have happened since you set up this PR I would suggest either starting from scratch on new branch or resolving all the conflicts

anmarques commented:
This PR is ready but needs to be merged by hand. Will create a new PR soon and delete this one when done.

anmarques commented:
Too much time has passed. Will re-implement if needed
